docs/architecture/adr-11-refactor-errors.md

```diff
@@ -0,0 +1,348 @@
+is known, along with the instance that was actually found. These are both included in the
+error under the `expected` and `actual` fields. The error message for this class should
+always include "expected ``{expected}``, actual ``{actual}``"; this statement should be precluded
+by "mismatched [TYPE]:". The type should be spelled in the singular and followed by a colon.
+The `expected` and `actual` interpolated values in the error message should be surrounded by
+``{}`` backticks and then braces.
+
+The rest of the error classes are a bit more freeform in how they are structured. They could
+take the form of unit structs that do not contain any values, serving mainly to surface a
+specific error message and nothing more.
+
+```rust
+    /// missing attribute key
+    MissingAttributeKey,
+    /// missing attribute value
+    MissingAttributeValue,
+```
+
+These two variants each highlight the fact that a very particular resource is missing.
+
+Single-element unit structs are used primarily to surface an invalid or duplicate type,
+opting to not provide any additional context other than what the error message itself
+provides.
+
+```rust
+    /// invalid status `{0}`
+    InvalidStatus(Status),
+    /// duplicate client state `{0}`
+    DuplicateClientState(ClientId),
+```
+
+The most common form of variant is a one-element struct with a `description` field. This
+serves to allow injecting more context at the call site of the error in the form of a
+String. This is the most general variant: care should be taken to not allow these sorts
+of variants to be abused for unintended purposes. Note that, like variants that contain
+nested errors, descriptions should be interpolated in the error message without enclosing
+backticks.
+
+```rust
+    /// failed to verify header: {description}
+    FailedToVerifyHeader { description: String },
+    /// invalid hash: {description}
+    InvalidHash { description: String },
+```
+
+## Future Work
+
+In light of the stated downsides, a natural follow-up of this work would be to generalize
+ibc-rs's error handling architecture to allow hosts to introduce their own bespoke error
+types. This would allow host developers to define more precise error types and variants
+that can better signal what the root cause of an error might be. This would improve upon
+the rather generic error variants that are exposed through ibc-rs's own `HostError` definition.
+
+The downside is that this would add additional work on top of the already considerable amount 
+of work that it takes to integrate ibc-rs into host chains.
+
+[dispatch]: https://github.com/cosmos/ibc-rs/blob/4aecaece9bda3c0f4a3b6a8379d73bd7eddc2cc4/ibc-core/ics25-handler/src/entrypoint.rs#L35
+[validate]: https://github.com/cosmos/ibc-rs/blob/4aecaece9bda3c0f4a3b6a8379d73bd7eddc2cc4/ibc-core/ics25-handler/src/entrypoint.rs#L54
+[execute]: https://github.com/cosmos/ibc-rs/blob/4aecaece9bda3c0f4a3b6a8379d73bd7eddc2cc4/ibc-core/ics25-handler/src/entrypoint.rs#L130
```

Comment:
It would be great to link the relevant issues/PRs in a `Reference` section to highlight the steps have been taken.